// action-core/esm5/index.js
import { __assign, __awaiter, __generator, __rest } from "tslib";
/**
 * JSON type
 * https://github.com/microsoft/TypeScript/issues/27930
 */
import { JSON } from 'ts-brand-json';
var ActionCore = /** @class */ (function () {
    function ActionCore() {
        this.handles = {};
    }
    ActionCore.prototype.install = function (type, actionTrigger) {
        Object.defineProperty(this.handles, type, {
            value: actionTrigger,
        });
    };
    ActionCore.prototype.uninstall = function (type) {
        Object.defineProperty(this.handles, type, {
            value: void 0,
        });
        delete this.handles[type];
    };
    ActionCore.prototype.run = function (action) {
        return __awaiter(this, void 0, void 0, function () {
            var type, param, data, target, option, handle, paramData, _a, _target, _b, urlParams, urlData;
            return __generator(this, function (_c) {
                type = action.type, param = action.param, data = action.data, target = action.target, option = __rest(action, ["type", "param", "data", "target"]);
                handle = this.handles[type];
                paramData = {};
                if (param) {
                    try {
                        paramData = JSON.parse(param);
                    }
                    catch (error) {
                        console.error("[Action-Core Error]: action param parse error \u2193\n", error);
                    }
                }
                _a = target.split('?'), _target = _a[0], _b = _a[1], urlParams = _b === void 0 ? '' : _b;
                urlData = {};
                urlParams.split('&').map(function (v) {
                    var _a = v.split('='), key = _a[0], value = _a[1];
                    urlData[key] = value;
                });
                return [2 /*return*/, handle({
                        target: _target,
                        data: __assign(__assign(__assign(__assign({}, paramData), data), option), urlData),
                    })];
            });
        });
    };
    return ActionCore;
}());
export default ActionCore;
